espnet2/speechlm/core_lm/valle.py

```diff
@@ -0,0 +1,315 @@
+from espnet2.speechlm.module.module import (
+    TransformerLayer,
+    LevelAwareTransformerLayer,
+    PositionalEncoding,
+)
+from espnet2.speechlm.net_utils import (
+    length_mask,
+    ce_loss,
+    install_kv_cache_hook,
+    logits_to_tokens,
+)
+
+
+class ValleLM(AbsCoreLM):
+    def __init__(
+        self,
+        vocab_size: int,
+        nq: int,
+        share_emb: bool = True,
+        pos_enc: str = None,
+        att_unit: int = 256,
+        head: int = 2,
+        unit: int = 1024,
+        ar_layer: int = 4,
+        nar_layer: int = 4,
+        dropout_rate: float = 0.0,
+        positional_dropout_rate: float = 0.0,
+        attention_dropout_rate: float = 0.0,
+    ):
+        super(ValleLM, self).__init__()
+
+        if pos_enc == "sinusoidal":
+            pos_enc_class = PositionalEncoding
+        else:
+            raise ValueError(f"unknown pos-enc option: {pos_enc}")
+
+        self.emb = torch.nn.Embedding(vocab_size, att_unit)
+        self.lm_head = torch.nn.Linear(att_unit, vocab_size, bias=False)
+        if share_emb:
+            self.lm_head.weight = self.emb.weight
+
+        # Auto-Regressive part
+        self.ar_decoders = torch.nn.ModuleList(
+            [
+                TransformerLayer(
+                    att_unit=att_unit,
+                    head=head,
+                    unit=unit,
+                    dropout_rate=dropout_rate,
+                    attention_dropout_rate=attention_dropout_rate,
+                    causal=True,
+                    cross_attention=False,
+                )
+                for _ in range(ar_layer)
+            ]
+        )
+        self.ar_post_ln = torch.nn.LayerNorm(att_unit)
+        self.ar_pos_enc = pos_enc_class(att_unit, positional_dropout_rate)
+
+        # Non-Auto-Regressive part
+        self.nar_decoders = torch.nn.ModuleList(
+            [
+                LevelAwareTransformerLayer(
+                    att_unit=att_unit,
+                    head=head,
+                    unit=unit,
+                    dropout_rate=dropout_rate,
+                    attention_dropout_rate=attention_dropout_rate,
+                    causal=False,
+                    cross_attention=False,
+                    n_level=nq - 1,
+                )
+                for _ in range(nar_layer)
+            ]
+        )
+        self.nar_post_ln = torch.nn.LayerNorm(att_unit)
+        self.nar_pos_enc = pos_enc_class(att_unit, positional_dropout_rate)
+
+        self.nq = nq
+
+    def forward(
+        self,
+        dec_seq: torch.Tensor,
+        dec_seq_lengths: torch.Tensor = None,
+        enc_seq: torch.Tensor = None,
+        enc_seq_lengths: torch.Tensor = None,
+        prefix_len: torch.Tensor = None,
+    ) -> Tuple[torch.Tensor, torch.Tensor, Dict]:
+
+        assert dec_seq.dim() == 3
+
+        # Auto-Regressive part
+        input_ar = dec_seq[:, :-1, 0]
+        target_ar = dec_seq[:, 1:, 0]
```

Comment:
Please use docstring to unified the shape information. It is not always good to ask users to infer the shape everytime (especially when the index selection happens often)